model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ *
+ *
+ */
+
+namespace oat\taoQtiItem\model\qti\CustomInteractionAsset;
+
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\BaseExtendedCustomInteractionAssetExtractor;
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\TextReaderExtendedAssetExtractor;
+use oat\taoQtiItem\model\qti\interaction\CustomInteraction;
+use oat\taoQtiItem\model\qti\interaction\ImsPortableCustomInteraction;
+use oat\taoQtiItem\model\qti\interaction\PortableCustomInteraction;
+
+/**
+ * @author Kiryl Poyu <kyril.poyu@taotesting.com>
+ */
+class ExtendedCustomInteractionAssetExtractorAllocator
```

Comment:
This looks more like an abstract factory to me, WDYT?

I mean, it clearly _creates_ an extractor service.

Reply:
I do not think so. In general, this class is used just to determine which implementation of the base class we need. An abstract factory can be used as part of this flow if we have a complex process for creating descendants.

Reply:
LGTM, you just need to understand where this mapping will come from

Reply:
I suppose we can do that on the service provider level.
https://github.com/oat-sa/generis/blob/master/core/DependencyInjection/ContainerServiceProviderInterface.php

Reply:
probably in this case extractorMap should contain factories set  with some interface
```php
interface CustomInteractioAssetInstractorFactoryInterface {
   public function create(CustomInteraction $interaction): AssetExtractorInterface
}
```
and allocator in this case will be look like 
```php
public function __constuct(private array $extractorMap) {
}

public function allocateExtractor(CustomInteraction $interaction) {
    return $this->extractorMap[$interaction->getTypeIdentifier()]->create(interaction) ?? new NullAssetExtractor();
}
```
to keep flexibility in creation of some complex extractors

Reply:
Or that map itself can be a collection of callbacks, instantiating a required extractor.